[kafka-adapter] Add e2e error metrics -- to be continued

MMirelli opened the pull request:
Currently, the kafka adapter doensn't support neither e2e latency metrics nor e2e error metrics. These are all defined in the pulsar adapter: https://github.com/nosqlbench/nosqlbench/blob/12f8697e0e27754074eca413451c0707d0663d70/adapter-pulsar/src/main/java/io/nosqlbench/adapter/pulsar/util/PulsarAdapterMetrics.java#L42-L59.
This aims at integrating the e2e error metrics.

This contains the continuation of https://github.com/nosqlbench/nosqlbench/pull/1135.

eolivelli commented on adapter-kafka/src/main/java/io/nosqlbench/adapter/kafka/ops/OpTimeTrackKafkaConsumer.java:
```diff
@@ -191,6 +209,16 @@ private void updateE2ELatencyMetric(ConsumerRecord<String, String> record) {
     }
 
+    private void checkAndUpdateMessageErrorCounter(ConsumerRecord<String, String> record) {
+        String msgSeqIdStr = new String(record.headers().lastHeader(PulsarAdapterUtil.MSG_SEQUENCE_NUMBER).value());
```
what happens if there is no such header ?
do we see here a null or does it throw an exception ?

MMirelli replied:
In case there is not such a key among the header keys, a NPE would be thrown, as `value()` is applied on `null`.

eolivelli replied:
so we should handle that case (null) the same way as a blank string

MMirelli replied:
Now, I skip the metrics generation in case it's null and warn the user about it.

eolivelli commented on adapter-kafka/pom.xml:
```diff
@@ -76,6 +76,12 @@
+        <dependency>
+            <groupId>io.nosqlbench</groupId>
+            <artifactId>adapter-pulsar</artifactId>
+            <version>5.17.2-SNAPSHOT</version>
```
this should be ${project.version)

MMirelli replied:
good catch, thank you!

eolivelli commented on adapter-kafka/pom.xml:
```diff
@@ -76,6 +76,12 @@
             <version>2.8.0</version>
         </dependency>
+        <dependency>
+            <groupId>io.nosqlbench</groupId>
```
${project.groupId}

eolivelli commented on adapter-kafka/pom.xml:
```diff
@@ -76,6 +76,12 @@
         </dependency>
+        <dependency>
+            <groupId>io.nosqlbench</groupId>
+            <artifactId>adapter-pulsar</artifactId>
```
please add a comment that explains why we need this dependency

MMirelli replied:
https://github.com/nosqlbench/nosqlbench/pull/1143#issuecomment-1450375869

eolivelli replied:
I didn't mean to add a comment on the PR but to also drop a line like:

```
    <!-- This dependency is needed in order to re-use some utilities from the Pulsar adapter code -->
```

MMirelli replied:
Oh ok fixed.

MMirelli replied:
@jshook @yabinmeng - I didn't give enough context about this dep. 

I was suggesting to create a new module as a shared util module (generating metrics tracking message duplicates, message loss and re-ordered messages) as this could be easily reused by some module other than `adapter-pulsar` and `adapter-kafka`. For example a module which might need that in the future is `adapter-s4j`.

MMirelli replied:
I decoupled moved the reused code in `adapters-api` module in `io.nosqlbench.engine.api.metrics.ReceivedMessageSequenceTracker`. Also see: https://github.com/nosqlbench/nosqlbench/pull/1143/commits/b33389824980c41c7d227a465ffa5875287763d1.

eolivelli reviewed (commented):
overall LGTM
I have left a couple of comments

MMirelli commented:
In order to reuse the main util tracking the end-to-end errors (https://github.com/nosqlbench/nosqlbench/pull/1143/files#diff-80f671d2680a92638d696e350f048425a61a168840f63bf698dbcdb42e39141bR212-R221) I had to add `pulsar-adapter` as a dependency of the kafka adapter: https://github.com/nosqlbench/nosqlbench/pull/1143/files#diff-f6388beadc327908ba7cf42521a8083a1697d956f144edf7f25d1d4a4047a9c5R79-R84. 

In general, I think this is a bad practice. @jshook, may I fix this a bit later? I would like to first have this working in an internal testing env.

yabinmeng reviewed (changes requested):
LGTM overall with some comments. I'm not really a fan of adding dependency among NB modules.

MMirelli commented on adapter-kafka/src/main/java/io/nosqlbench/adapter/kafka/ops/OpTimeTrackKafkaProducer.java:
```diff
@@ -193,6 +206,11 @@ void cycleMsgProcess(long cycle, Object cycleObj) {
         }
 
         ProducerRecord<String, String> message = (ProducerRecord<String, String>) cycleObj;
+        if (seqTracking) {
```
I think this is ok for now. We can modify this later on, if needed.

MMirelli commented:
I also: 
1. tested that I didn't break the pulsar adapter metrics by using https://github.com/nosqlbench/nosqlbench/pull/1143/commits/8eb9498fa98944941d118d507dc732cb71832749 and https://github.com/nosqlbench/nosqlbench/pull/1143/commits/0a2ab1dcca00f1dab3409f4956cb71965e7322c7
2. fixed a change which was breaking e2e error metrics in the pulsar adapter: https://github.com/nosqlbench/nosqlbench/pull/1143/commits/971c5d3c6f6bfec5ebf1d103e6d936b19646c347
3. documented the document level e2e options: https://github.com/nosqlbench/nosqlbench/pull/1143/commits/7d5ecf2f32bd93e58d79d6b5279869ff28fc6662
4. added handy scripts and configuration files to quickly build and test out the adapters: https://github.com/nosqlbench/nosqlbench/pull/1143/commits/8eb9498fa98944941d118d507dc732cb71832749 and https://github.com/nosqlbench/nosqlbench/pull/1143/commits/0a2ab1dcca00f1dab3409f4956cb71965e7322c7

@jshook and @yabinmeng, would it be fine to keep the commits in 4?

MMirelli commented:
>@MMirelli I'm assuming this is still in the works, since it is still in "changes requested" state.

Unless you have more comments, I think I am done with this.

yabinmeng reviewed (commented):
I would recommend to seek Shooky's review since made some changes in the engine code.

MMirelli commented:
@jshook I would merge this, it's been fairly stable with up to 20K msg/s with transaction enabled.
